### PR TITLE
rebar3: refactor to make hermetic

### PR DESCRIPTION
**Pull request by ericbmerritt**
This adds changes to the rebar3 expression that patch rebar3 to force it
to be hermetic. Now, by default, rebar3 literally can't download
anything. A 'rebar3-open' expression was added for those folks whe want
the normal rebar3.

### CODE REVIEW

File: pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
++++ b/src/rebar_pkg_resource.erl
+@@ -100,10 +100,10 @@ make_vsn(_) ->
+     {error, "Replacing version of type pkg not supported."}.
+ 
+ request(Url, ETag) ->
+-    case httpc:request(get, {Url, [{"if-none-match", ETag} || ETag =/= false]},
+-                       [{ssl, ssl_opts(Url)}, {relaxed, true}],
+-                       [{body_format, binary}],
+-                       rebar) of
++    case rebar_hermicity:request(get, {Url, [{"if-none-match", ETag} || ETag =/= false]},
```
Comment (gleber) [inline pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch]: Is this module on purpose non-existent?

File: pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch [hunk 2]
```diff
@@ -0,0 +1,76 @@
+diff --git a/src/rebar3.erl b/src/rebar3.erl
+index 2b73844..455a5d5 100644
+--- a/src/rebar3.erl
++++ b/src/rebar3.erl
+@@ -65,6 +65,7 @@ main(Args) ->
+ 
+ %% Erlang-API entry point
+ run(BaseState, Commands) ->
++    start_and_load_apps(command_line),
```
Comment (gleber) [inline pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch]: Why this change?

File: pkgs/development/erlang-modules/hex-packages.nix [hunk 3]
```diff
@@ -1796,13 +1739,15 @@ let
 
     jc_1_0_4 = callPackage
       (
-        {  buildHex }:
+        {  buildHex, jsone_1_2_0, jwalk_1_1_0, lager_3_0_1, ranch }:
```
Comment (gleber) [inline pkgs/development/erlang-modules/hex-packages.nix]: These changes in this form will be a waste a bit, since they'll get overwritten with next updated of this file. We either need to do some sort of list of overrides in hex2nix or a fixed point overwrite mechanism here, which would survive update of the base file.

Comment (ericbmerritt) [inline pkgs/development/erlang-modules/hex-packages.nix]: Yup. they are a complete waste. I just wanted to get things working with minimal effort. I agree, I am just not sure what thats going to look like. Maybe a file of 'fixes' that we add over time that hex2nix picks up automatically

File: pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch (hunk 2 above)
Comment (ericbmerritt) [inline pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch]: Its an artifact of some additional logging I wanted to do. I thought I had dropped it. I will update the commit.

File: pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch (hunk 1 above)
Comment (ericbmerritt) [inline pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch]: no, it should be there. hmm. let me figure out why.

File: pkgs/development/erlang-modules/hex-packages.nix (hunk 3 above)
Comment (gleber) [inline pkgs/development/erlang-modules/hex-packages.nix]: Ack.

**Comment by ericbmerritt** [comment]
@gleber this is ready to be pushed.

File: pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch (hunk 2 above)
Comment (gleber) [inline pkgs/development/tools/build-managers/rebar3/hermetic-rebar3.patch]: This is still in the latest PR code

**Comment by gleber** [comment]
```diff
+    start_and_load_apps(command_line),
```

Once this is gone, feel free to merge.

**Comment by ericbmerritt** [comment]
@gleber this is good.
